libtenzir/builtins/operators/decapsulate.cpp

```diff
@@ -268,109 +281,110 @@ auto parse(record_ref builder, std::span<const std::byte> bytes,
   return std::nullopt;
 }
 
-struct operator_args {
-  std::optional<located<uint16_t>> vxlan_port;
-
-  friend auto inspect(auto& f, operator_args& x) -> bool {
-    return f.object(x)
-      .pretty_name("operator_args")
-      .fields(f.field("vxlan_port", x.vxlan_port));
+auto decapsulate(series s, diagnostic_handler& dh) -> std::optional<series> {
+  if (s.type.name() != "pcap.packet") {
+    diagnostic::warning("cannot decapsulate schema '{}'", s.type.name())
+      .note("schema must be 'pcap.packet'")
+      .emit(dh);
+    return std::nullopt;
   }
-};
+  // Get the packet payload.
+  const auto& layout = caf::get<record_type>(s.type);
+  const auto linktype_index = layout.resolve_key("linktype");
+  if (!linktype_index) {
+    diagnostic::warning("got a malformed 'pcap.packet' event")
+      .note("schema 'pcap.packet' must have a 'linktype' field")
+      .emit(dh);
+    return std::nullopt;
+  }
+  // TODO: ADD CHECK
+  auto linktype_array
+    = linktype_index->get(caf::get<arrow::StructArray>(*s.array));
+  auto linktype_values = caf::get_if<arrow::UInt64Array>(&*linktype_array);
+  if (!linktype_values) {
+    diagnostic::warning("got a malformed 'pcap.packet' event")
+      .note("field 'linktype' not of type uint64")
+      .emit(dh);
+    return std::nullopt;
+  }
+  const auto data_index = layout.resolve_key("data");
+  if (!data_index) {
+    diagnostic::warning("got a malformed 'pcap.packet' event")
+      .note("schema 'pcap.packet' must have a 'data' field")
+      .emit(dh);
+    return std::nullopt;
+  }
+  auto data_array = data_index->get(caf::get<arrow::StructArray>(*s.array));
+  auto data_values = caf::get_if<arrow::BinaryArray>(&*data_array);
+  if (!data_values) {
+    diagnostic::warning("got a malformed 'pcap.packet' event")
+      .note("field 'data' not of type blob")
+      .emit(dh);
+    return std::nullopt;
+  }
+  auto builder = series_builder{};
+  for (auto i = 0u; i < s.length(); ++i) {
+    auto linktype = (*linktype_values)[i];
+    auto data = (*data_values)[i];
+    if (!data) {
+      continue;
+    }
+    auto raw_frame = std::span<const std::byte>{
+      reinterpret_cast<const std::byte*>(data->data()), data->size()};
+    auto inferred_type = static_cast<frame_type>(linktype ? *linktype : 0);
+    if (auto diag = parse(builder.record(), raw_frame, inferred_type)) {
+      dh.emit(std::move(*diag));
+    }
+  }
+  // Add back the untouched data column at the end.
```

Comment:
This part might be specific to the operator version. The function probably does not need it.

Reply:
Could you elaborate a little on why is it not needed?

Reply:
Sure. Lets say I use `decapsulate` as a function: `y = decapsulate(x)`. By returning "the untouched data column" as `pcap` in the record, you now `y.pcap` and `x`, which refer to exactly the same data, which is kinda unnecessary.